usersession/userd/privileged_desktop_launcher.go

```diff
@@ -0,0 +1,249 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package userd
+
+import (
+	"bufio"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+
+	"github.com/godbus/dbus"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/strutil/shlex"
+)
+
+const privilegedLauncherIntrospectionXML = `
+<interface name="org.freedesktop.DBus.Peer">
+	<method name='Ping'>
+	</method>
+	<method name='GetMachineId'>
+               <arg type='s' name='machine_uuid' direction='out'/>
+	</method>
+</interface>
+<interface name='io.snapcraft.PrivilegedDesktopLauncher'>
+	<method name='OpenDesktopEntry'>
+		<arg type='s' name='desktopFileID' direction='in'/>
+	</method>
+</interface>`
+
+// PrivilegedDesktopLauncher implements the 'io.snapcraft.PrivilegedDesktopLauncher' DBus interface.
+type PrivilegedDesktopLauncher struct {
+	conn *dbus.Conn
+}
+
+// Name returns the name of the interface this object implements
+func (s *PrivilegedDesktopLauncher) Interface() string {
+	return "io.snapcraft.PrivilegedDesktopLauncher"
+}
+
+// BasePath returns the base path of the object
+func (s *PrivilegedDesktopLauncher) ObjectPath() dbus.ObjectPath {
+	return "/io/snapcraft/Launcher"
+}
+
+// IntrospectionData gives the XML formatted introspection description
+// of the DBus service.
+func (s *PrivilegedDesktopLauncher) IntrospectionData() string {
+	return privilegedLauncherIntrospectionXML
+}
+
+// OpenDesktopEntry implements the 'OpenDesktopEntry' method of the 'io.snapcraft.DesktopLauncher'
+// DBus interface. The desktopFileID is described here:
+// https://standards.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html#desktop-file-id
+func (s *PrivilegedDesktopLauncher) OpenDesktopEntry(desktopFileID string, sender dbus.Sender) *dbus.Error {
+	desktopFile, err := desktopFileIDToFilename(desktopFileID)
+	if err != nil {
+		return dbus.MakeFailedError(err)
+	}
+
+	err = verifyDesktopFileLocation(desktopFile)
+	if err != nil {
+		return dbus.MakeFailedError(err)
```

Comment:
by construction atm we don't expect to ever get there because desktopFileIDToFilename can only produce results that pass verifyDesktopFileLocation ? is that right?  in that case it seems a comment would be good, and also adding "internal error:" to the error here

Reply:
here would be fine, it's really about the interaction of the two functions called here

Reply:
I think **here** is the wrong place for a comment: it is likely to go unchanged and silently become misleading.

In the hypothetical scenario where we enable launching from non-snap desktop ids, the _code_ here won't need any change and neither compiler nor tests will highlight the comment.

Reply:
well, indeed the problem we have at the moment is that I can make verifyDesktopFileLocation() just return nil and no test fails, because of its nature unit tests for verifyDesktopFileLocation() itself might help but not a lot.

Personally I tend to read the call code of single usage functions before changing them, but that might be me.

Reply:
> I think this is overall fairly close, but there are some missing unit tests. I have not reviewed the spread test yet either, but I'm happy to help with the unit tests, let me know if you need help with that.

I'm not sure when I'll have the headspace to get back into this, certainly not this week. So anything you can offer to move things forward would be appreciated.

Reply:
ack, I'll split some of these things out into smaller pr's and hopefully make this one a bit smaller/easier to review and land